Version 2.0

RomanDubovskyi opened the pull request:
Added:
 - tests using Junit 4
Upgraded:
 - pom.xml (removed unneeded property)
 - getLongestWord, getSecondLongestWord (optimized implementation)
 - removed static methods
 - removed redundant instructions from Readme file
 - used try with resources

RomanDubovskyi committed:
Version 2.0
Added:
 - tests using Junit 4
Upgraded:
 - pom.xml (removed unneeded property)
 - getLongestWord, getSecondLongestWord (optimized implementation)
 - removed static methods
 - removed redundant instructions from Readme file
 - used try with resources

boroda4436 commented on src/main/java/Solution.java:
```diff
@@ -4,33 +4,37 @@
-        String secondLongestWord = concatWords.get(concatWords.size() - 2);
-        printResultValues(concatenatedWordsAmount, longestConcatenatedWord, secondLongestWord);
+        concatWords.remove(Collections.max(concatWords, Comparator.comparing(String::length)));
+        return Collections.max(concatWords, Comparator.comparing(String::length));
```
виходить так, що ти двічі викликаєш collection.max
Це не сильно б'є по перформансу?

RomanDubovskyi replied:
оптимізував цей метод, тепер використовую лише один for each loop

boroda4436 commented on src/main/java/Solution.java:
```diff
@@ -4,33 +4,37 @@
     }
 
-    public static List<String> getConcatenatedWords(List<String> words) {
+    private List<String> getConcatenatedWords(List<String> words) {
```
Я б зробив цей метод default та покрив за допомогою тестів

RomanDubovskyi replied:
не знаю, чи я цей коментар правильно виправив, бо раніше не мав справи з дефолтними методами на практиці
Я створий інтефейс, де цей і ще два (колишніх приватних методи) зробив дефолтними і  зробив свій клас Solution implements DataReadFromFile.
Якщо я зробив це все ок, можеш пояснити чим в цьмоу випадку таке рішення оптимальніше ніж те, що було раніше і коли загалом такий підхід застсовувати
можливо якусь статтю по цьому порадиш ?

RomanDubovskyi replied:
тестами, дефолтні методи я покрив

boroda4436 replied:
та ні) малось на увазі модифікатор видимості default замість private) Все набагато простіше)

boroda4436 commented on src/test/java/TestSolution.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Before
+    public void setUp() throws Exception {
+        solution = new Solution(new File("src/main/resources/words.txt"));
```
оце вже виглядає круто)

boroda4436 replied:
я б для тестів використав окремий файл, який би лежав у папці test/resources

RomanDubovskyi replied:
Додав окремий файл file_for_testing.txt для тестів виключно, який наочно дублює приклад з умови до тестового завдання. 
Не знаю наскільки це доречно, проте, враховуючи, що в тестах ми не отримуємо результатів з того файлу, що прикрплений до завдання. + щоб додати трішки більше кастомності аплікусі я додав ще Main клас, де можна консольно через менюху працювати з апкою і просто вводити назву файлу (наприклад "words") і отримувати по ньому різну інфу, яку вимагає завдаання...

boroda4436 replied:
all good) зазвичай коли ти пишеш якусь консольну аплікуху яка буде працювати з файлами - то на вхід передаєш шлях до файлу

boroda4436 commented on src/test/java/TestSolution.java:
```diff
@@ -0,0 +1,45 @@
+    @Test(expected = RuntimeException.class)
+    public void checkUsingInvalidFile() {
+        solution = new Solution(new File("src/main/resources/wrongfile.txt"));
+        assertEquals(14024, solution.getAmountOfConcatWords());
```
Якщо ти ловиш runtime exception, то чи доходить твій тест до цього рядку?

RomanDubovskyi replied:
Доходив
проте я всеодно підправив реалізацію, щоб було чіткіше видно, на якому куску коду, я це очікую

RomanDubovskyi committed:
Version 3.0
Added:
 - interface DataReadFromFile with default methods
 - tests for default methods
 - util class Menu, Main
 - file_for_testing.txt (to test app on simple example)
Upgraded:
 - getSecondLongestWord (optimized implementation)
 - readme file (added instruction on how to run app)

boroda4436 reviewed (approved):
Good job)
